### ISSUE TITLE
SSH Agent Forwarding not working with Remote SSH on MacOS (still a problem)

### ISSUE THREAD

**Issue by mbmccoy** [microsoft/vscode#168202]

**Comment by tm1000** [comment]
This is now happening from my coworker (@gamegamer43) and myself since the latest update. It seems like the commands:

```
[14:36:09.935] Updating $SSH_AUTH_SOCK: ln -f -s "/tmp/ssh-XXXXwXtrPD/agent.4194" "/run/user/0/vscode-ssh-auth-sock-934342336"
[14:36:09.937] Remote extension host environment: {"SSH_AUTH_SOCK":"/run/user/0/vscode-ssh-auth-sock-934342336"}
```

Are not actually being run. When. I run the two above manually the key is correctly passed.

**Comment by jt-oh** [comment]
This is also happening in windows 10.

As refered as [previous comment](https://github.com/microsoft/vscode/issues/168202#issuecomment-1419882326), below command was logged but not executed.

```
[10:27:28.317] Updating $SSH_AUTH_SOCK: ln -f -s "/tmp/ssh-TQUU6UJGJL/agent.11754" "/run/user/1000/vscode-ssh-auth-sock-952306072"
```

The symbolic link was not created on the destination.

```
remote$ ls -a /run/user/1000/vscode-ssh-auth-sock-952306072
ls: cannot access '/run/user/1000/vscode-ssh-auth-sock-952306072': No such file or directory
```

And ssh-agent in local is not forwarded to remote.

```
remote$ ssh-add -l
Error connecting to agent: No such file or directory
```

After run the below command, symbolic link was created and ssh-agent was forwarded successfully.

```
remote$ ln -f -s "/tmp/ssh-TQUU6UJGJL/agent.11754" "/run/user/1000/vscode-ssh-auth-sock-952306072"
remote$ ls -a /run/user/1000/vscode-ssh-auth-sock-952306072
/run/user/1000/vscode-ssh-auth-sock-952306072
remote$ ssh-add -l
2048 SHA256:L7TcPUCaZ************* \****\***\.ssh\*** (RSA)
```

Below is My VSCode version

```
Version: 1.75.0 (user setup)
Commit: e2816fe719a4026ffa1ee0189dc89bdfdbafb164
Date: 2023-02-01T15:23:45.584Z
Electron: 19.1.9
Chromium: 102.0.5005.194
Node.js: 16.14.2
V8: 10.2.154.23-electron.0
OS: Windows_NT x64 10.0.19045
Sandboxed: No
```

**Comment by gonzigonz** [comment]
Same thing happing to me. I found that disabling, reloading, and enabling `Remote.SSH: Use Local Server` works for me but given this looks like a race condition issue, I may just have been lucky.

**Comment by madalinignisca** [comment]
Fresh installation of MacOS Ventura, added keys to ssh-agent, installed VSCode, reusing my synced ssh config file (identical same one from my Debian Linux laptop) and ssh forwarding not working on the Mac. Works on my Linux laptop although. Nothing special on that machine, as reinstalled about 3 weeks ago, with all working, no issues.

**Comment by nourspace** [comment]
> Fresh installation of MacOS Ventura, added keys to ssh-agent, installed VSCode, reusing my synced ssh config file (identical same one from my Debian Linux laptop) and ssh forwarding not working on the Mac. Works on my Linux laptop although. Nothing special on that machine, as reinstalled about 3 weeks ago, with all working, no issues.
> 
> Killing vscode server on the remote machine and reconnecting fixed the issue. I think that vscode server running on the remote machine is expecting the old agent connection.
> 
> Is there a way to make the running process exit when vscode disconnects?

Great tip, thanks! I had to remove the binaries `rm -rf ~/.vscode-server/bin` for this to work

**Comment by orbsmiv** [comment]
Having used agent forwarding without issue for a few years this has suddenly stopped working for me.

At a glance it seems that following the initial connection (where agent forwarding works), subsequent connections to the SSH server are using an outdated symlink to the agent in `/tmp`. I've tried a few of the suggestions listed above but with no success as yet.

**Comment by madalinignisca** [comment]
@orbsmiv VSCode has a command to kill all code servers on the current ssh connection. Try that. Worked for me, and I just tried it between 2 machines with the issue above being possible to replicate it.

**Comment by maaarghk** [comment]
@madalinignisca I just tested in a working remote session: I used the `Remote-SSH: Kill VS Code server on host` and also `Remote-SSH: Kill current VS code server` commands, and found that when I made a new connection to that server, I still experience the race condition where the extension attempted to run the `ln` command potentially before registering a terminal backend.  There was a low % chance that it would work without intervention, but mostly I had to continue to use the workaround where the `ln` command from the Output window is manually copied to a terminal. This was true whether the remote session attempted to reload previous terminal sessions or not.

**Comment by davidkhess** [comment]
Dropping a reference to @maaarghk ’s debugging of this issue on another issue that was just closed without resolution:

https://github.com/microsoft/vscode-remote-release/issues/7995#issuecomment-1422753238

**Comment by linhnvhicts** [comment]
> > Fresh installation of MacOS Ventura, added keys to ssh-agent, installed VSCode, reusing my synced ssh config file (identical same one from my Debian Linux laptop) and ssh forwarding not working on the Mac. Works on my Linux laptop although. Nothing special on that machine, as reinstalled about 3 weeks ago, with all working, no issues.
> > Killing vscode server on the remote machine and reconnecting fixed the issue. I think that vscode server running on the remote machine is expecting the old agent connection.
> > Is there a way to make the running process exit when vscode disconnects?
> 
> Great tip @madalinignisca , thanks! I had to remove the binaries `rm -rf ~/.vscode-server/bin` for this to work

confirmed this is working if you remake the agents on your machine before ssh, because `/.vscode-server/bin` save the old agents

**Comment by davidkhess** [comment]
@mbmccoy you may not be paying much attention to this issue any longer, but can you edit the title of it to say: "SSH Agent Forwarding sometimes not working with Remote SSH"?

I think that might help get more attention to it and help keep it from getting accidentally closed.

**Comment by madalinignisca** [comment]
> I am having this issue too. My ssh agent is not getting forwarded with vscode remote, but it is correctly forwarded if I just do normal ssh via a terminal. In my case, agent forwarding in vscode works for some remotes, but fails for remotes for which I have a `ProxyJump` setting in the ssh config file. Could the `ProxyJump` setting cause some particular issues?

I doubt. I have my setup with a proxy jump.

**Comment by northluke** [comment]
Still having this issue, ssh on a standalone terminal works fine but vscode integrated terminal doesn't set SSH_AUTH_SOCK environment variable. Simply setting the variable manually makes the integrated terminal work until some amount of idle time and I have to do it again.

**Comment by aastafiev** [comment]
Still having issue after upgrade to 1.76.0

**Comment by aastafiev** [comment]
No. It’s not related 


---------------------
Sincerely yours,
Aleksey Astafiev

**Comment by gonzigonz** [comment]
> Same thing happing to me. I found that disabling, reloading, and enabling `Remote.SSH: Use Local Server` works for me but, given this looks like a race condition issue, I may just have been lucky.

I kept doing this for a few days after that post, and the issue still occurred but was a lot less frequent. So I then disabled `Remote.SSH: Use Local Server` altogether and haven't had issues since.

**Comment by aastafiev** [comment]
> 

Thank you. Have changed and it works yet.

**Comment by tm1000** [comment]
> I kept doing this for a few days after that post, and the issue still occurred but was a lot less frequent. So I then disabled `Remote.SSH: Use Local Server` altogether and haven't had issues since.

This also worked for me! 🤷

**Comment by madalinignisca** [comment]
> > Same thing happing to me. I found that disabling, reloading, and enabling `Remote.SSH: Use Local Server` works for me but, given this looks like a race condition issue, I may just have been lucky.
> 
> I kept doing this for a few days after that post, and the issue still occurred but was a lot less frequent. So I then disabled `Remote.SSH: Use Local Server` altogether and haven't had issues since.

@gonzigonz That means that every time you reconnect "fresh", a new vscode server process will be started for you. There should be a new feature to ask the user to kill the other processes.

I find it normal for vscode server to "die" when there is no other connection to it, but... it survives.

If not, in some time you'll wonder why vscode processes use all server ram.

**Comment by alpsayin** [comment]
> > Same thing happing to me. I found that disabling, reloading, and enabling `Remote.SSH: Use Local Server` works for me but, given this looks like a race condition issue, I may just have been lucky.
> 
> I kept doing this for a few days after that post, and the issue still occurred but was a lot less frequent. So I then disabled `Remote.SSH: Use Local Server` altogether and haven't had issues since.

I don't know why but this also worked for me. After restarting both vscode (not just reload) and vscode-server that is.
Linux client:Linux remote. Will try Windows client: Linux remote as well.

**Comment by gonzigonz** [comment]
> @gonzigonz That means that every time you reconnect "fresh", a new vscode server process will be started for you.  

@madalinignisca  Are you sure? A new vscode server process is started locally or on the remote machine?

> If not, in some time you'll wonder why vscode processes use all server ram.  

Just to be clear, I haven't root-caused the issue. I just found this workaround and have yet to notice any performance degradation. This isn't to say there aren't any, but this would be difficult to assess since many use cases and factors exist. FWIW, I work on a reasonably large repo every day. I also started using SSH `control master`  a while back, as I found it speeded up the time it took to create new integrated terminals.

**Comment by madalinignisca** [comment]
@gonzigonz I run a server for all team to use it with ssh/docker, and I did notice that some have hanging vscode server processes hanging even for days there once they ticked the Remote.SSH: Use Local Server. I ended up to perform a kill every night.

**Comment by alpsayin** [comment]
Disabling "use local server" did not work for Windows client (1.76.2) and Rhel8 host. 
I ended up crafting the below code and injecting it into my `.bashrc`
```
if [[ "$SSH_AUTH_SOCK" =~ "vscode" ]]; then
  # echo "This is a vscode-terminal. SSH_AUTH_SOCK=$SSH_AUTH_SOCK"
  ssh-add -l 1>/dev/null 2>/dev/null
  if [[ "$?" != "0" ]]; then
    export SSH_AUTH_SOCK=$(ls -lt /tmp/ssh-**/* | awk '{print $9}' | head -n1 | grep -Po '.*\d+');
    # echo "Replaced SSH_AUTH_SOCK to $SSH_AUTH_SOCK"
  fi
fi
```

**Comment by austin-seyboldt** [comment]
Can we get an update on this? I've never had an issue before with remote-ssh and for the last month it has been very flakey. Sometimes killing the remote server or local server works. Sometimes I need to restart my computer. I have ssh forwarding configured properly and it works from a regular terminal, just not vscode. I noticed this started happening for me when I upgrade to 1.76. I downgraded to 1.75 and that fixed it but this morning it started happening again (still on 1.75).

**Comment by tm1000** [comment]
> Disabling "use local server" did not work for Windows client (1.76.2) and Rhel8 host. I ended up crafting the below code and injecting it into my `.bashrc`
> 
> ```
> if [[ "$SSH_AUTH_SOCK" =~ "vscode" ]]; then
>   # echo "This is a vscode-terminal. SSH_AUTH_SOCK=$SSH_AUTH_SOCK"
>   ssh-add -l 1>/dev/null 2>/dev/null
>   if [[ "$?" != "0" ]]; then
>     export SSH_AUTH_SOCK=$(ls -lt /tmp/ssh-**/* | awk '{print $9}' | head -n1 | grep -Po '.*\d+');
>     # echo "Replaced SSH_AUTH_SOCK to $SSH_AUTH_SOCK"
>   fi
> fi
> ```

This worked for me. I had to stop disabling useLocalServer because of what @madalinignisca stated.

**Comment by patroza** [comment]
None of the tricks work for me, especially while signing git commits with ssh as well.

**Comment by thibaut-lo** [comment]
I made a shim script that I source in my .zshrc.

It is just a variation of other solutions posted above, I only added a check that the last vscode ssh sock is a valid symlink.

```bash
#!/usr/bin/env bash

fix_vscode_ssh() {
  if [ -e "$SSH_AUTH_SOCK" ] && [ -h "$SSH_AUTH_SOCK" ]; then # Valid symlink
    echo Valid symlink for SSH_AUTH_SOCK $SSH_AUTH_SOCK
    return 0;
  fi

  echo Invalid symlink for SSH_AUTH_SOCK: $SSH_AUTH_SOCK

  local redefine_alias=false
  if alias ls >/dev/null 2>&1; then # Existing alias
    unalias ls # exa is printing where the symlink redirects while I just want the path of the symlink
    redefine_alias=true 
  fi

  local new_ssh=$(ls -t  /run/user/${UID}/vscode-ssh-auth-*  | head -1)

  if [ redefine_alias ]; then
    alias ls="exa --color=auto --icons" # TBC depending on your own alias settings
  fi

  if ! ([ -e "$new_ssh" ] && [ -h "$new_ssh" ]); then # Invalid symlink
    printf "Latest vscode symlink [%s] for ssh is also dead. Open a new vsode workspace manually.\n" $new_ssh;
    return 1
  fi

  export SSH_AUTH_SOCK=$new_ssh

  echo SSH_AUTH_SOCK set to $SSH_AUTH_SOCK.

  return 0
}

if [ "$TERM_PROGRAM" = "vscode" ]; then
  fix_vscode_ssh
fi
```
and then in my `.zshrc`

`source <path_of_the_above_script.sh>`

**Comment by jonathansnell** [comment]
I can confirm that this still an issue, and that it is not specific to Mac, as we have both Mac and Linux users affected by it. On the working servers it shows the `"Update SSH_AUTH_SOCK" terminal command done` as usual, and on the non-working servers it does not. We are loading from workspace files on the remote hosts.

**Comment by iraadit** [comment]
It is an issue for me too.
None of the previous mentioned workarounds seem to work

**Comment by patroza** [comment]
in my current setup, if I have ssh local server enabled, I connect, and it doesn't work, I can run: "Restart extension host", and then it starts working. I might need to re-open a terminal if it was already open.

**Comment by iraadit** [comment]
Workaround that finally worked for me:
```
SSH_AUTH_SOCK=$(ls -td /tmp/ssh**/* | head -n1)
ssh-add -l
```

**Comment by sdaguet** [comment]
Also facing the same issue, workaround is also working.
Just wondering why VSCode server creates a symlink in `/run/user/$uid/... to /tmp/ssh*`

`[14:20:49.988] Updating $SSH_AUTH_SOCK: ln -f -s "/tmp/ssh-G2w8LvqiuZ/agent.281639" "/run/user/1001/vscode-ssh-auth-sock-420156547"`

Sounds the link is not created. From Output console: 
`ln -f -s "/tmp/ssh-zP7cLxc6yg/agent.282444" "/run/user/1001/vscode-ssh-auth-sock-956100014"`

Then : 
```
$ ls "/run/user/1001/vscode-ssh-auth-sock-956100014"
: ... '/run/user/1001/vscode-ssh-auth-sock-956100014': No such file or directory
```

If I create the link manually, all works well.

Why not define $SSH_AUTH_SOCK to /tmp/..., or fix the link creation issue ?

**Comment by skwde** [comment]
Same issue, see also what I posted here <https://github.com/microsoft/vscode/issues/168325#issuecomment-1449819219>.

Probably the issues are related.

**Comment by jon-dez** [comment]
restarting the host system fixes this for me, so it is most likely a vscode server side issue

**Comment by 3f6a** [comment]
@jon-dez The "fix" is only temporary, I presume?

**Comment by jon-dez** [comment]
> @jon-dez The "fix" is only temporary, I presume?
> 
> 
> 
> I also see that the SSH key is correctly forwarded sometimes, but after a while it stops working again.

Yea, I should have mentioned it was a temp fix

**Comment by sdaguet** [comment]
Hello,

Usually I solve it using the following:

1. Check output / Remote-ssh
2. Find Updating $SSH_AUTH_SOCK: ln -f -s "/tmp/ssh-byTiqsbv9T/agent.xxxx" "/run/user/1001/vscode-ssh-auth-sock-yyyy"
3. Execute the above command

To get our ducks in a raw, my issue occurs when I connect to a remote ssh server directly, and want to establish manually from terminal some ssh connections to other remote hosts using my local ssh key, which is for sure not deployed on remote servers.

**Comment by prosenboim** [comment]
Any progress on this? Another VS code update and git is still broken

**Comment by tommasodiotalevi** [comment]
Following this thread. 
I am also experiencing the same issue, trying to perform git operations in a remote connection vscode shell, with the ForwardAgent option set to True on my ssh_config file (MacOS 13.4.1). 
In a separate shell, not on vscode, I can perfectly execute this command `ssh -T git@github.com` with correct authn but not in the plugin terminal. From the log files, the vscode terminal is looking for the wrong pubkey (not looking at the forwarded key).

**Comment by jmskov** [comment]
Just chiming in, I have encountered this a few times now. Works for months, and suddenly the stale SSH behavior pops up. On 1.82.0 locally. My steps to reproduce:
1. Open remote with "Remote-SSH: Connect to Host..." to a server with fresh code server
2. In new window, try `ssh -T git@github.com`, successful auth
3. Add seemingly any new folder to the workspace
4. Try authentication command again, fails this time
5. Delete the code server `~/.vscode-server`
6. Go to 1.

**Comment by bdittmer** [comment]
> Just chiming in, I have encountered this a few times now (including today, right now lol). Works for months, and suddenly the stale SSH behavior pops up. On 1.82.0 locally. My steps to reproduce:
> 
> 1. Open remote with "Remote-SSH: Connect to Host..." to a server with fresh code server
> 2. In new window, try `ssh -T git@github.com`, successful auth
> 3. Add seemingly any new folder to the workspace
> 4. Try authentication command again, fails this time
> 5. Delete the code server `~/.vscode-server`
> 6. Go to 1.
> 
> Edit: to be clear, using Agent forwarding from the host that has worked previously

I was able to get agent forwarding working again by  setting `remote.SSH.useExecServer: false`. It looks like the issue you and I are experiencing was reported here: https://github.com/microsoft/vscode-remote-release/issues/8926

**Comment by jmskov** [comment]
@bdittmer Thanks for pointing this out. I want to add that I was experiencing the SSH bug I described before updating to 1.82.0 as well (I updated to 1.82.0 trying to squash the original issue).

**Comment by akutruff** [comment]
I'm experiencing a flavor this issue on Windows with dev containers.  The interesting this is that Debian based containers work totally fine, but when I use the ubuntu base image, I can clone the repository with ssh but thereafter all git operations stop working.  

Under the Ubuntu container,  `ssh -T git@github.com` fails due to ` Permission denied (publickey).`  However if I do `ssh-add -l` in the same terminal it prints the correct public key.  

I am using `mcr.microsoft.com/vscode/devcontainers/base:ubuntu`

**Comment by rademacher-p** [comment]
I'm also experiencing this x11 forwarding issue from Local (Windows) > SSH (Linux) > dev container.

- Local > dev container: ✔️ 
- Local > SSH > `docker run` (or `docker compose`): ✔️ 
- Local > SSH > dev container: ❌ 

Note that the local > dev container forwarding works with zero `devcontainer.json` configuration for environment vars, bind mounts, etc. (as needed for success w/ `docker run`)

Got the SSH forwarding, GPU, etc. all straight - this is the last thing holding me back from fully containerizing my development 😅 Let me know if there's any additional info I can provide to help debug!

**Comment by zadorski** [comment]
Same here, fresh installation of VSCode on Windows is connecting to Ubuntu Server host (fresh installation).
 
Regular terminal (Windows Terminal PS session):
- `ssh -vT git@github.com` is authenticated successfully.

Integrated terminal (VSCode Remote SSH session):
- `ssh-add -l` shows forwarded key fine. 
- `ssh -vT git@github.com` ends with dreaded "git@github.com: Permission denied (publickey)".

I tried lots of permutations at this point.

Steps:
1. `$SSH_AUTH_SOCK` didn't match to the current agent at the remote session
2. made it straight with the suggested workaround: 
`export SSH_AUTH_SOCK=$(ls -t /tmp/ssh-**/* | head -1)`
3. however, the same `Permission denied`

Context: 
- `terminal.integrated.inheritEnv` and `remote.SSH.enableAgentForwarding` are both enabled in VSCode.
- .ssh config is trivial
   ```
   Host 192.168.1.100
       ForwardAgent yes
   ```

```
Version: 1.83.1 (user setup)
Commit: f1b07bd25dfad64b0167beb15359ae573aecd2cc
Date: 2023-10-10T23:48:05.904Z
Electron: 25.8.4
ElectronBuildId: 24154031
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Windows_NT x64 10.0.22621
```

**Comment by zadorski** [comment]
With workaround it worked on Windows with Ubuntu Server SSH host.

Extra touches:
- `remote.SSH.useLocalServer` left TRUE, as it hadn't influenced the issue
- `remote.SSH.remoteServerListenOnSocket` set to TRUE

Steps (in addition to details above): 
1. Updated OpenSSH on the Windows host (from v8.6 to v9.4):
    `winget install Microsoft.OpenSSH.Beta`     
2. Exported `$SSH_AUTH_SOCK` in the integrated terminal within remote SSH session: 
    `export SSH_AUTH_SOCK=$(ls -td /tmp/ssh**/* | head -n1)` 
    
Just for the reference:
`$SSH_AUTH_SOCK` was `/run/user/1000/vscode-ssh-auth-sock-*` upon connection to remote host.
`$SSH_AUTH_SOCK` became `/tmp/ssh-XXXXBHX4WS/agent.*` after step 2 applied.

**Comment by IvanCobotic** [comment]
I'm running on windows 11, working against a remote host on Linux. Ive had this issue happen to me occasionally. My symtom is exactly as the original poster, 
ssh to remote in a  WSL (ubuntu) terminal.
`ssh -T git@github.com`  works fine
From a vscode terminal i get "permision denied"
running 
`$ export SSH_AUTH_SOCK=$(ls -t /tmp/ssh-**/* | head -1)`
in the vscode terminal fixes the issue and after that github agent works fine.

**Comment by karolpiczak** [comment]
> I was able to get agent forwarding working again by setting `remote.SSH.useExecServer: false`. It looks like the issue you and I are experiencing was reported here: [microsoft/vscode-remote-release#8926](https://github.com/microsoft/vscode-remote-release/issues/8926)

+1 on this. 

In my case using the `ls -t /tmp/ssh-**/* | head -1` workaround successfully addresses the issue for terminals. However, when `useExecServer` is set to `true`, there is still an unresolved problem with the built-in git functionality, such as syncing changes from the source control view. It appears that the `useExecServer` implementation may not be properly handling the `SSH_AUTH_SOCK` changes.

**Comment by albireox** [comment]
Having the same issue on MacOS using fish shell and connecting to Linux remote servers over ssh.

For me the `useExecSever: false` doesn't solve the problem. Overriding the `$SSH_AUTH_SOCK` with a valid socket works in the integrated terminal, but apart from being a hack it doesn't solve the problem of getting the integrated Git features to work on a remote machine.

**Comment by pascal456** [comment]
just opened https://github.com/microsoft/vscode-remote-release/issues/9393

The problem only occurs for me when connection is established through remote tunnels (remote network). AgentForwarding via remote ssh is fine (local network). Then again when opening a Dev Container via the remote tunnels connection, the forwarding is working.

Any chance that this is related?

**Comment by albireox** [comment]
Any news on this? This problem is still happening and I am not able to get a persistent agent with any of the newer stable or insiders versions of VSCode.

**Comment by mbmccoy** [comment]
I hit this again, apparently triggered by logging from a terminal outside of VSCode. Previous workarounds failed, including restarting the terminal. 

### Current workaround

1. SSH into machine _outside of VSCode_, but forwarding remote agent: `ssh -A me@my-machine`. 
2. In this terminal, echo the $SSH_AUTH_SOCK variable: 
    ```console
    echo $SSH_AUTH_SOCK
    ```
    which will produce something like
    ```
    /tmp/ssh-dHPZ4apwPl/agent.1100176
    ```
3. In the VSCode integrated terminal, set the environment variable:

        export SSH_AUTH_SOCK=<from-above>

**Comment by d-sauter** [comment]
i had this issue maybe a year ago, cant remember what i did to "permanently" fix it. but this week it seemingly randomly appeared. 

I vaguely remember adding some extra flag for when vscode starts up the shell.

its annoying having to run `export SSH_AUTH_SOCK=$(ls -t /tmp/ssh-**/* | head -1)` all the time

**Comment by Jud1cator** [comment]
> its annoying having to run `export SSH_AUTH_SOCK=$(ls -t /tmp/ssh-**/* | head -1)` all the time

Just encountered the same problem, it seems you can add `export SSH_AUTH_SOCK=$(ls -t /tmp/ssh-**/* | head -1)` to your `.bashrc`/`.zshrc`/whatever to avoid being annoyed :smile:

**Comment by gopipalamalai** [comment]
This issue seems to have started again today. vscode can't seem to fix this issue. Keeps coming back every year. see https://github.com/microsoft/vscode/issues/175806

**Comment by lesca** [comment]
> [](https://github.com/microsoft/vscode-remote-release/issues/8926)

You are awesome! You saved my day!

**Comment by lesca** [comment]
> > Just chiming in, I have encountered this a few times now (including today, right now lol). Works for months, and suddenly the stale SSH behavior pops up. On 1.82.0 locally. My steps to reproduce:
> > 
> > 1. Open remote with "Remote-SSH: Connect to Host..." to a server with fresh code server
> > 2. In new window, try `ssh -T git@github.com`, successful auth
> > 3. Add seemingly any new folder to the workspace
> > 4. Try authentication command again, fails this time
> > 5. Delete the code server `~/.vscode-server`
> > 6. Go to 1.
> > 
> > Edit: to be clear, using Agent forwarding from the host that has worked previously
> 
> I was able to get agent forwarding working again by setting `remote.SSH.useExecServer: false`. It looks like the issue you and I are experiencing was reported here: [microsoft/vscode-remote-release#8926](https://github.com/microsoft/vscode-remote-release/issues/8926)

You are awesome!

+1 on this
In my case, agent is forwarding with `ssh` in cmd (this means config file is set correctly).
But in VS Code, agent not forwarding, until I changed this setting `remote.SSH.useExecServer` to false.

**Comment by ferreteleco** [comment]
Any update on this, it is getting very hard to work on development containers from windows.

**Comment by ascii78** [comment]
I've tried logging in with an old 1.6 vscode i had on a usb stick with static config. After logging in remote-ssh tells me the following symlinks are once again established:

[13:58:44.035] Updating $SSH_AUTH_SOCK: ln -f -s "/tmp/ssh-XXXXiuO5un/agent.203985" "/run/user/1000/vscode-ssh-auth-sock-956309600"
[13:58:44.035] Using cwd: vscode-remote://ssh-remote%2Bhostname/
[13:58:44.035] Remote extension host environment: {"SSH_AUTH_SOCK":"/run/user/1000/vscode-ssh-auth-sock-956309600"}

cleaning up all the remote.SSH.* settings and workarounds on 1.8 and logging in again with that version the symlinks started working again.

**Comment by haplav** [comment]
> Was experiencing this issue today. Can confirm that `"remote.SSH.useExecServer": false,` fixes it.

Same here right now!

FWIW,
```
>Remote-SSH: Uninstall VS Code Server from Host...
```
just before connecting works for me as well...

**Comment by micahcarroll** [comment]
Just in case anyone else has this issue, and `"remote.SSH.useExecServer": false` does not fix it for them:

Deleting `~/.vscode-server` fixed it for me

**Comment by ehoogeveen-medweb** [comment]
I started having this issue on Windows as well. Previously, closing and reopening the terminal helped, but that stopped working. Setting `remote.SSH.useExecServer` to `false` and restarting VSCode seems to have helped for now.

**Comment by david-boles** [comment]
I just experienced this on Windows as well:

```
Version: 1.89.1 (user setup)
Commit: dc96b837cf6bb4af9cd736aa3af08cf8279f7685
Date: 2024-05-07T05:13:33.891Z
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Windows_NT x64 10.0.19045

Remote-SSH version - v0.110.1
```

`SSH_AUTH_SOCK` was set to a path that didn't exist in new terminals. Manually setting `SSH_AUTH_SOCK` to the path in the `Remote - SSH` output worked fine. Disabling `remote.SSH.useExecServer` also seems to have worked.

**Comment by sbonds** [comment]
Deleting the remote `$HOME/.vscode-server` and letting it auto-install AFTER enabling ForwardAgent seems to have resolved the issue for me as well. It seems like if that initial connection is made without ForwardAgent set, future connections get stuck in that config even when ForwardAgent is enabled later.

**Comment by floroz** [comment]
I can confirm these steps solved it for me:

- remove forwarding from local/remote ssh configs (comment out)
- close all VSCode instances
- on the remote, delete `/home/your-user/.vscode-server`
- re-enable the AgentForwarding in both configs
- relaunch VSCode, it will re-install the VSCode server
- VSCode terminal should now have SSH forwarding enabled

**Comment by mabdulker** [comment]
Both solutions worked for me:
- [Solution 1](https://github.com/microsoft/vscode/issues/168202#issuecomment-1773794916)
- [Solution 2](https://github.com/microsoft/vscode-remote-release/issues/9662#issuecomment-2010515888)

**Comment by kylebfredrickson** [comment]
Setting `remote.SSH.useExecServer` seemed to work initially but then failed. The only thing that reliably fixes the problem for me is to delete `.vscode-server` in home.

The problem changes depending on how `remote.SSH.useExecServer` is set though.
- if set, when echoing SSH_AUTH_SOCK I get a stale `/tmp/ssh-*/agent.*`
- if not set, I get a stale `/run/user/1000/vscode-ssh-auth-sock-*`.

**Comment by tweakyTweeter** [comment]
I also faced the same issue on MacOS. I was able to fix it by deleting ~/.vscode-server on remote machine and then reconnecting via Remote SSH extension.

**Comment by lobantseff** [comment]
The setting set which seems to work for my setup (macOS client -> Ubuntu host):

```json
{
    "remote.SSH.enableAgentForwarding": true,
    "remote.SSH.useLocalServer": false,
    "remote.SSH.useExecServer": false,
    "remote.SSH.remoteServerListenOnSocket": true,
}
```

**Comment by loodvn** [comment]
Thanks @lobantseff, works for me! I'll update here if anything changes on my side though. 
Could you please include in your comment some reasoning behind why it works, for posterity?

**Comment by PhoenixmitX** [comment]
> The setting set which seems to work for my setup (macOS client -> Ubuntu host):
> 
> ```json
> {
>     "remote.SSH.enableAgentForwarding": true,
>     "remote.SSH.useLocalServer": false,
>     "remote.SSH.useExecServer": false,
>     "remote.SSH.remoteServerListenOnSocket": true,
> }
> ```

also fixes ubuntu -> ubuntu

**Comment by mach10x** [comment]
The workarounds mentioned here stops working for me (Windows client-> Linux Host) when I 'Reload Window'.
For me setting export SSH_AUTH_SOCK=$(ls -t /tmp/ssh-**/* | head -1) on .zshrc works.
To make that command run only in VSCode integrated terminal, I did it like https://stackoverflow.com/questions/45635168/vscode-how-to-run-a-command-after-each-terminal-open

**Comment by leohku** [comment]
> The setting set which seems to work for my setup (macOS client -> Ubuntu host):
> 
> ```json
> {
>     "remote.SSH.enableAgentForwarding": true,
>     "remote.SSH.useLocalServer": false,
>     "remote.SSH.useExecServer": false,
>     "remote.SSH.remoteServerListenOnSocket": true,
> }
> ```

Apart from these commands, I also had to add
```json
{
    "terminal.integrated.persistentSessionReviveProcess": "never",
    "terminal.integrated.enablePersistentSessions": false,
}
```
To prevent the stale terminal session from being restored. After that, agent forwarding worked flawlessly.

**Comment by Tyriar** [comment]
@joshspicer since you're very familiar with the SSH extension now, could you look into this issue and come up with recommendations to fix? Do we need to special case certain terminals to not revive the processes or something?

**Comment by dmpopoff** [comment]
> Was experiencing this issue today. Can confirm that `"remote.SSH.useExecServer": false,` fixes it.

This works for me as well. But in my case (VSCode on Windows 11 23H2, cmd.exe -> ssh -V 8.2) additionally I also needed to update the ssh client. I updated to OpenSSH_for_Windows_9.8p1 Win32-OpenSSH-GitHub, LibreSSL 3.9.2.

options in my local ssh_config for this remote Linux host:
```
  IdentityFile C:\\Users\\user1\\.ssh\\id_ed25519
  ForwardAgent yes
  AddKeysToAgent yes
```

**Comment by joshspicer** [comment]
Apologies for no follow up.  We've been tracking (and have hopefully fixed) these agent forwarding issues in the latest pre-release of the extension.  

The code has now changed many times since this issue was created, so i'll close it as a dupe of https://github.com/microsoft/vscode-remote-release/issues/10274
